### PR TITLE
Add multiple options to fundrawtransaction

### PR DESCRIPTION
**Pull request by promag**
This PR allows to specify an address to receive the change of the fund. Currently `fundrawtransaction` accepts a boolean for the second parameter. With this change the second parameter can be either a boolean or a JSON object with the following optional keys: `includeWatching`, `changeAddress` and `changePosition`.

### CODE REVIEW

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -2468,7 +2468,7 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
                             + HelpExampleCli("sendrawtransaction", "\"signedtransactionhex\"")
                             );
 
-    RPCTypeCheck(params, boost::assign::list_of(UniValue::VSTR)(UniValue::VBOOL));
+    // RPCTypeCheck(params, boost::assign::list_of(UniValue::VSTR)(UniValue::VBOOL));
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: Mixed feelings about how to do this since the second argument can be one of two types.

File: src/wallet/wallet.h [hunk 2]
```diff
@@ -735,7 +735,7 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
      * Insert additional inputs into the transaction by
      * calling CreateTransaction();
      */
-    bool FundTransaction(CMutableTransaction& tx, CAmount& nFeeRet, int& nChangePosRet, std::string& strFailReason, bool includeWatching);
+    bool FundTransaction(CMutableTransaction& tx, CAmount& nFeeRet, int& nChangePosRet, std::string& strFailReason, bool includeWatching, const CTxDestination& destChange);
```
Comment (promag) [inline src/wallet/wallet.h]: Don't know if the best place is at the end.

File: src/wallet/rpcwallet.cpp (hunk 1 above)
Comment (luke-jr) [inline src/wallet/rpcwallet.cpp]: IIRC you can just leave off the second type for RPCTypeCheck and it will skip it. Then do the check by hand.

**Comment by luke-jr** [comment]
You need to update the help. I suggest replacing the Boolean with the Object entirely, and just supporting the boolean in code as a backward compatibility thing.

File: src/wallet/wallet.h (hunk 2 above)
Comment (luke-jr) [inline src/wallet/wallet.h]: As good a place as any. But you need to update the implementation too...

Comment (promag) [inline src/wallet/wallet.h]: Ops, missed that in `git add -p`.

**Comment by promag** [comment]
@luke-jr done.

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -2468,7 +2472,7 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
                             + HelpExampleCli("sendrawtransaction", "\"signedtransactionhex\"")
                             );
 
-    RPCTypeCheck(params, boost::assign::list_of(UniValue::VSTR)(UniValue::VBOOL));
+    RPCTypeCheck(params, boost::assign::list_of(UniValue::VSTR)(UniValue::VOBJ));
```
Comment (luke-jr) [inline src/wallet/rpcwallet.cpp]: The VOBJ check should be independent of RPCTypeCheck if you want to support boolean for backward compatibility.

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -377,7 +377,7 @@ bool CWallet::Verify(const string& walletFile, string& warningString, string& er
         } catch (const boost::filesystem::filesystem_error&) {
             // failure is ok (well, not really, but it's not worse than what we started with)
         }
-        
+
```
Comment (luke-jr) [inline src/wallet/wallet.cpp]: Please don't change unrelated whitespace like this.

File: src/wallet/rpcwallet.cpp (hunk 3 above)
Comment (promag) [inline src/wallet/rpcwallet.cpp]: I see, so the help says it accepts a JSON object, but the implementation allows both?

**Comment by promag** [comment]
@luke-jr ok, added a test for the changeAddress.

File: src/wallet/rpcwallet.cpp [hunk 5]
```diff
@@ -2478,15 +2482,31 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
     if (origTx.vout.size() == 0)
         throw JSONRPCError(RPC_INVALID_PARAMETER, "TX must have at least one output");
 
+    CBitcoinAddress changeAddress;
```
Comment (luke-jr) [inline src/wallet/rpcwallet.cpp]: This probably needs to be defaulting to CNoDestination()

File: src/wallet/wallet.h (hunk 2 above)
Comment (luke-jr) [inline src/wallet/wallet.h]: A default CNoDestination() here might be a good idea.

**Comment by luke-jr** [comment]
Also, this conflicts with #7159 , so one of them will need to rebase. I think #7159 should adopt the options Object used here.

**Comment by promag** [comment]
@luke-jr agree 7159 should adopt options.

**Comment by promag** [comment]
@luke-jr done.

**Comment by promag** [comment]
Added changePosition option to specify the desired vout index of change address.

File: qa/rpc-tests/fundrawtransaction.py [hunk 6]
```diff
@@ -575,7 +601,7 @@ def run_test(self):
         outputs = {self.nodes[2].getnewaddress() : watchonly_amount / 2}
         rawtx = self.nodes[3].createrawtransaction(inputs, outputs)
 
-        result = self.nodes[3].fundrawtransaction(rawtx, True)
+        result = self.nodes[3].fundrawtransaction(rawtx, {'includeWatching': True })
```
Comment (MarcoFalke) [inline qa/rpc-tests/fundrawtransaction.py]: Can you leave one of them and mention that this is testing backward compatibility?

Comment (promag) [inline qa/rpc-tests/fundrawtransaction.py]: :+1:

File: qa/rpc-tests/fundrawtransaction.py [hunk 7]
```diff
@@ -591,6 +617,7 @@ def run_test(self):
         outputs = {self.nodes[2].getnewaddress() : watchonly_amount}
         rawtx = self.nodes[3].createrawtransaction(inputs, outputs)
 
+        # Backward compability test (2nd param is includeWatching)
```
Comment (promag) [inline qa/rpc-tests/fundrawtransaction.py]: @MarcoFalke FYI

File: qa/rpc-tests/fundrawtransaction.py [hunk 8]
```diff
@@ -591,6 +617,7 @@ def run_test(self):
         outputs = {self.nodes[2].getnewaddress() : watchonly_amount}
         rawtx = self.nodes[3].createrawtransaction(inputs, outputs)
 
+        # Backward compatibility test (2nd param is includeWatching)
         result = self.nodes[3].fundrawtransaction(rawtx, True)
```
Comment (promag) [inline qa/rpc-tests/fundrawtransaction.py]: @MarcoFalke FYI.

File: src/wallet/wallet.h [hunk 9]
```diff
@@ -735,7 +735,7 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
      * Insert additional inputs into the transaction by
      * calling CreateTransaction();
      */
-    bool FundTransaction(CMutableTransaction& tx, CAmount& nFeeRet, int& nChangePosRet, std::string& strFailReason, bool includeWatching);
```
Comment (promag) [inline src/wallet/wallet.h]: I feel we should have a better argument order. Suggestion:
```cxx
bool FundTransaction(CMutableTransaction& tx, const CTxDestination& changeDestination, int& changePosition, bool includeWatching, CAmount& nFeeRet, std::string& strFailReason);
```

Comment (luke-jr) [inline src/wallet/wallet.h]: It doesn't really matter as long as existing code either fails to compile or behaves the same. Note that to put destChange earlier, you would need to make it a required option.

File: src/wallet/rpcwallet.cpp [hunk 10]
```diff
@@ -2468,7 +2473,31 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
...
 
-    RPCTypeCheck(params, boost::assign::list_of(UniValue::VSTR)(UniValue::VBOOL));
+    RPCTypeCheck(params, boost::assign::list_of(UniValue::VSTR));
+
+    CTxDestination changeAddress = CNoDestination();
+    int changePosition = -1;
+    bool includeWatching = false;
+
+    if (params.size() > 1) {
+      if (params[1].type() == UniValue::VBOOL) {
```
Comment (jonasschnelli) [inline src/wallet/rpcwallet.cpp]: If we go for the backward compatibility way: I guess then this code part should be commented as "backward compatibility bool only fallback".

File: src/wallet/wallet.h (hunk 9 above)
Comment (jonasschnelli) [inline src/wallet/wallet.h]: I guess we should rename `nChangePosRet ` to `nChangePosInOut` and add a comment in the header that -1 as a input value will result in setting a random position.

File: src/wallet/wallet.cpp [hunk 11]
```diff
@@ -2134,14 +2136,17 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
                     // add the dust to the fee.
                     if (newTxOut.IsDust(::minRelayTxFee))
                     {
+                      nChangePos = -1;
```
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: nit: indent.

File: src/wallet/wallet.cpp [hunk 12]
```diff
@@ -2134,14 +2136,17 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
...
                         nFeeRet += nChange;
                         reservekey.ReturnKey();
                     }
                     else
                     {
-                        // Insert change txn at random position:
-                        nChangePosRet = GetRandInt(txNew.vout.size()+1);
-                        vector<CTxOut>::iterator position = txNew.vout.begin()+nChangePosRet;
+                        if (nChangePos == -1) {
+                          // Insert change txn at random position:
```
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: nit: indent (2 instead 4), maybe run the clang format script.

File: src/wallet/rpcwallet.cpp [hunk 13]
```diff
@@ -2468,7 +2473,31 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
...
+        includeWatching = params[1].get_bool();
+      }
+      else {
+        RPCTypeCheck(params, boost::assign::list_of(UniValue::VSTR)(UniValue::VOBJ));
+
+        UniValue options = params[1];
+
+        if (options.exists("changeAddress"))
```
Comment (jonasschnelli) [inline src/wallet/rpcwallet.cpp]: nit: Not sure if we should make this check case insensitive. Very likely, some users will try it with "changeaddress".

File: src/wallet/rpcwallet.cpp [hunk 14]
```diff
@@ -2468,7 +2473,31 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
...
+
+    if (params.size() > 1) {
+      if (params[1].type() == UniValue::VBOOL) {
+        includeWatching = params[1].get_bool();
+      }
+      else {
+        RPCTypeCheck(params, boost::assign::list_of(UniValue::VSTR)(UniValue::VOBJ));
+
+        UniValue options = params[1];
+
```
Comment (jonasschnelli) [inline src/wallet/rpcwallet.cpp]: Maybe add a `RPCTypeCheckObj()` to ensure that `changePosition`, etc. are the correct types? Otherwise you get a plain `JSON value is not a boolean as expected` (etc.)?

**Comment by jonasschnelli** [comment]
utACK.
Nice work!

I'm not sure if we should add the backward compatibility fallback (includeWatchonly = true). Because it could also lead to problems in future. If so, we should also mention in in the RPC help (something like `a plain True results in includeWatchonly=true`).

IMO a API break would be acceptable.

File: src/wallet/rpcwallet.cpp (hunk 13 above)
Comment (luke-jr) [inline src/wallet/rpcwallet.cpp]: I think we already have precedent for case-sensitive options?

**Comment by nunofgs** [comment]
@promag: it would be useful to add an option "lockUnspents: true" to ensure that chosen unspents are effectively locked before sending the transaction.

**Comment by promag** [comment]
@nunofgs yes, that idea was also in #7498.

File: src/wallet/rpcwallet.cpp (hunk 14 above)
Comment (promag) [inline src/wallet/rpcwallet.cpp]: :+1:

**Comment by promag** [comment]
@nunofgs lockUnspents option added. Missing test for the moment.

**Comment by promag** [comment]
Visited all comments and rebased. Please review.

**Comment by promag** [comment]
What about adding unlockUnspents option to `sendrawtransaction`?

**Comment by ruimarinho** [comment]
@promag I think that should be part of the standard process - as soon as the transaction is sent to the network, the unspents should be unlocked and considered spent.

**Comment by jonasschnelli** [comment]
@ruimarinho: you probably mean "locking". I think locking of the unspents should happen somewhere between sign and send. I'm fine with an option in fundrawtransaction, but IMO we should not make it the default.

**Comment by promag** [comment]
@sipa @laanwj do you think it's acceptable to specify a fixed fee and bypass the minimal fee calculation through iteration in `CWallet::CreateTransaction`?

**Comment by promag** [comment]
@jonasschnelli sure, default for locking is false. The use case is to prevent consecutive `fundrawtransactions` calls selecting the same unspents.

**Comment by ruimarinho** [comment]
@jonasschnelli if I call `lockUnspent(false, [unspents])` manually, those unspents will remain unlocked until bitcoind is restarted or `lockUnspent(true, [unspents])` is called. I did not see a code path where these unspents are cleaned up - removed from the "locked" list but still unavailable for future spending.

**Comment by jonasschnelli** [comment]
@ruimarinho: Yes. There is no such thing as a auto-release because the intention is the keep these unspents locked.

But if you lock the unspents in `fundrawtransaction` by default, you need to unlock them in case you don't sign or send them (or in cases of an error).
But I agree, there are use-cases where you want to directly lock the unspent in the fundrawtransaction call (concurrency).

**Comment by promag** [comment]
@jonasschnelli the same applies if someone locks some unspents with `lockunspent` and don't use it.

**Comment by jonasschnelli** [comment]
@promag: yes. Agree. The default value is debatable.

**Comment by promag** [comment]
@laanwj @sipa bump.

**Comment by promag** [comment]
Rebased.

File: qa/rpc-tests/fundrawtransaction.py (hunk 6 above)
Comment (promag) [inline qa/rpc-tests/fundrawtransaction.py]: See below.

File: src/wallet/rpcwallet.cpp (hunk 14 above)
Comment (promag) [inline src/wallet/rpcwallet.cpp]: See below.

File: src/wallet/wallet.cpp [hunk 15]
```diff
@@ -1927,26 +1928,28 @@ bool CWallet::FundTransaction(CMutableTransaction& tx, CAmount &nFeeRet, int& nC
...
-        tx.vout.insert(tx.vout.begin() + nChangePosRet, wtx.vout[nChangePosRet]);
+    if (nChangePos != -1)
+        tx.vout.insert(tx.vout.begin() + nChangePos, wtx.vout[nChangePos]);
 
     // Add new txins (keeping original txin scriptSig/order)
     BOOST_FOREACH(const CTxIn& txin, wtx.vin)
     {
         if (!coinControl.IsSelected(txin.prevout))
             tx.vin.push_back(txin);
+        }
```
Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: Where does this scope start?

Comment (promag) [inline src/wallet/wallet.cpp]: It doesn't :)

**Comment by jonasschnelli** [comment]
I really think this is useful!
@promag can you look at the travis issues, maybe rebase it? I promise to test it afterwards. :)

File: src/wallet/wallet.cpp (hunk 15 above)
Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: Mind to fix it?

**Comment by promag** [comment]
@MarcoFalke fixed
@jonasschnelli rebased and renamed `nChangePosRet` to `nChangePosInOut`. I wonder if other `*Ret` should be renamed to `*Out`?

File: src/wallet/wallet.cpp [hunk 16]
```diff
@@ -1908,7 +1908,7 @@ bool CWallet::SelectCoins(const vector<COutput>& vAvailableCoins, const CAmount&
     return res;
 }
 
-bool CWallet::FundTransaction(CMutableTransaction& tx, CAmount &nFeeRet, int& nChangePosRet, std::string& strFailReason, bool includeWatching)
+bool CWallet::FundTransaction(CMutableTransaction& tx, CAmount &nFeeRet, int& nChangePosInOut, std::string& strFailReason, bool includeWatching, const CTxDestination& destChange)
```
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: missing `bool lockUnspents,` argument.

**Comment by jonasschnelli** [comment]
I think this PR should be in 0.13 (once finalized)!
Tested a bit after fixing the compile error (see above).

**Some issues:**

**1)**
I called

`fundrawtransaction 01000000000100e1f505000000001976a914f592d5ff0f4cad760d23601e32714639c244d27588ac00000000 '{"changeAddress":"mfrczJz4p5ziD4CiVTip3R7jjXJAjDKBmf", "includeWatching":true, "changePos":0}'
`
I used `changePos` but the valid argument would be `changePosition`. An error in a such case would be nice.

**2)**
Also, I used a invalid change address...
`fundrawtransaction 01000000000100e1f505000000001976a914f592d5ff0f4cad760d23601e32714639c244d27588ac00000000 '{"changeAddress":"dontforget.to.call@your.mother", "includeWatching":true, "changePosition":0}'
`
... which was accepted (no error or warning) but not used. It used one from the wallets pool (which could be dangerous in a watchOnly-fund-environment.

* [ ] detect invalid options and reject command with an error
* [ ] detect invalid change addresses and reject if so

Happy to test again.

**Comment by promag** [comment]
@jonasschnelli Thanks, I'll add tests to cover those cases. BTW, travis is not picking this PR?

**Comment by promag** [comment]
Please review.

@jonasschnelli FYI

 - [x] detect invalid options and reject command with an error: 7ba5713
 - [x] detect invalid change addresses and reject if so

**Comment by ruimarinho** [comment]
@jonasschnelli, since you're more involved in this PR, what do you think of @promag's suggestion above?

> do you think it's acceptable to specify a fixed fee and bypass the minimal fee calculation through iteration in CWallet::CreateTransaction?

**Comment by promag** [comment]
@ruimarinho maybe a separate PR on top of this because that is a change on `CWallet::CreateTransaction`.

**Comment by jonasschnelli** [comment]
@ruimarinho: Yes. A common problems of Bitcoin Core pulls is the "expanding scope". Keep it simple, add other features later in a different PR.

I think we need to distinct FRT and keep it – relatively simple. If we add options to set the fee, the only thing thats left is the coin selection then? not? Maybe someone could just grab the unspents over `listunspents` and do his own coin selection and create a final transaction with `createrawtransaction`.

And if you know the changeoutputs address, you could rewrite the transaction after FRT (on your side of the application) and change its fee.

@promag: thanks. Will re-test soon.

**Comment by promag** [comment]
> And if you know the changeoutputs address, you could rewrite the transaction after FRT (on your side of the application) and change its fee.

@jonasschnelli this is not entirely correct because there may be a better subset of unspents to better satisfy the output value + fixed fee. But since the fee is very low this may be a good tradeoff.

**Comment by promag** [comment]
> the only thing thats left is the coin selection then

@jonasschnelli actually this is accomplished with `createrawtransaction`.

**Comment by jonasschnelli** [comment]
> [...] And the worst case is when the total input value of the selected unspents isn't enough to cover the fixed fee. So I would say that having the option to specify the fixed fee is the way to go.

Yes. This is a point.

>> the only thing thats left is the coin selection then

> actually this is accomplished with createrawtransaction.

createrawtransaction does not use coinselection IMO. FRT does.

File: src/wallet/wallet.cpp [hunk 17]
```diff
@@ -1927,26 +1928,32 @@ bool CWallet::FundTransaction(CMutableTransaction& tx, CAmount &nFeeRet, int& nC
...
 
     // Add new txins (keeping original txin scriptSig/order)
     BOOST_FOREACH(const CTxIn& txin, wtx.vin)
     {
         if (!coinControl.IsSelected(txin.prevout))
+        {
             tx.vin.push_back(txin);
+
+            if (lockUnspents)
+                LockCoin(txin.prevout);
```
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: You need to hold `cs_wallet` for this operation.
I think between L1944 and L1945 you should add a `LOCK2(cs_main, cs_wallet);`

**Comment by jonasschnelli** [comment]
The type check works and I think it's great.
Missing LockCoin lock needs to be fixed.

**Comment by promag** [comment]
@jonasschnelli Thanks! Fixed and rebased.

**Comment by ruimarinho** [comment]
I was not suggesting otherwise regarding merging this PR and only then creating a new one to add the fixed fee feature :) like @promag mentioned on the edited comments, I believe there is still space for such feature as at least two scenarios are not covered by standard calls.

File: src/wallet/rpcwallet.cpp [hunk 18]
```diff
@@ -2451,8 +2451,14 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
...
                             "Only pay-to-pubkey, multisig, and P2SH versions thereof are currently supported for watch-only\n"
                             "\nArguments:\n"
-                            "1. \"hexstring\"     (string, required) The hex string of the raw transaction\n"
-                            "2. includeWatching (boolean, optional, default false) Also select inputs which are watch only\n"
+                            "1. \"hexstring\"           (string, required) The hex string of the raw transaction\n"
+                            "2. options               (object, optional)\n"
+                            "   {\n"
+                            "     \"changeAddress\"     (string, optional, default pool address) The bitcoin address to receive the change\n"
+                            "     \"changePosition\"    (numeric, optional, default random) The index of the change output\n"
+                            "     \"includeWatching\"   (boolean, optional, default false) Also select inputs which are watch only\n"
```
Comment (ruimarinho) [inline src/wallet/rpcwallet.cpp]: Bikeshedding but have you considered `includeWatchOnly` to follow a consistent nomenclature?

Comment (promag) [inline src/wallet/rpcwallet.cpp]: The underlying flag is allowWatchOnly. `includeWatching` is the old parameter name. Happy to rename if others agree.

Comment (jonasschnelli) [inline src/wallet/rpcwallet.cpp]: IMO we should keep `includeWatching`.

File: src/wallet/rpcwallet.cpp [hunk 19]
```diff
@@ -2451,8 +2451,14 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
                             "in the wallet using importaddress or addmultisigaddress (to calculate fees).\n"
                             "Only pay-to-pubkey, multisig, and P2SH versions thereof are currently supported for watch-only\n"
                             "\nArguments:\n"
-                            "1. \"hexstring\"     (string, required) The hex string of the raw transaction\n"
-                            "2. includeWatching (boolean, optional, default false) Also select inputs which are watch only\n"
+                            "1. \"hexstring\"           (string, required) The hex string of the raw transaction\n"
+                            "2. options               (object, optional)\n"
```
Comment (jonasschnelli) [inline src/wallet/rpcwallet.cpp]: nit: here we should mention something like : "for backward compatibility: passing in a `true` instead of an object will result in `{"includeWatching":true}"`

File: src/wallet/wallet.cpp [hunk 20]
```diff
@@ -1927,26 +1928,35 @@ bool CWallet::FundTransaction(CMutableTransaction& tx, CAmount &nFeeRet, int& nC
...
     CReserveKey reservekey(this);
     CWalletTx wtx;
-    if (!CreateTransaction(vecSend, wtx, reservekey, nFeeRet, nChangePosRet, strFailReason, &coinControl, false))
+    if (!CreateTransaction(vecSend, wtx, reservekey, nFeeRet, nChangePosInOut, strFailReason, &coinControl, false))
         return false;
 
-    if (nChangePosRet != -1)
-        tx.vout.insert(tx.vout.begin() + nChangePosRet, wtx.vout[nChangePosRet]);
+    if (nChangePosInOut != -1)
+        tx.vout.insert(tx.vout.begin() + nChangePosInOut, wtx.vout[nChangePosInOut]);
```
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: Here we need another check.
What if `nChangePosInOut = 10` but `tx.vout.size()==2`?
We need a out-of-range exception in a such case.

**Comment by jonasschnelli** [comment]
Allmost done. Last iteration:
* [ ] needs out-of-range detection of `changePosition`.
* [ ] needs simple documentation of backward compatibility (`true` instead of an parameter object)

**Comment by promag** [comment]
@jonasschnelli updated and rebased. not sure where is the best place to put the new help message.

File: src/wallet/wallet.cpp (hunk 20 above)
Comment (promag) [inline src/wallet/wallet.cpp]: Check is in rpcwallet.cpp.

Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: @promag: The check should should really be here (maybe an additional check). If someone uses `CWallet::FundTransaction()` with a `nChangePosInOut >= wtx.vout.size()`, the app crashes/throws an exceptions. Future changes might easy get trapped by this.

The problematic part is `wtx.vout[nChangePosInOut]`. There should be a check somewhere in `CWallet::FundTransaction()` that leads to a `return false` if a overflow has detected.

**Comment by jonasschnelli** [comment]
ping @promag: Would be nice if you could take the change output order check into `CreateTransaction()`: https://github.com/bitcoin/bitcoin/pull/7518/files#r58496250 
I'm convince that this is the last issue before this is ready for merging.

Please also fix the commit order.  The head commit is from March 5, but the HEAD~2 is March 30.
```
commit 4b0360f4160e059b61a21a95a6106d0eed927bd9
Author: João Barbosa <joao@uphold.com>
Date:   Sat Mar 5 09:45:57 2016 +0000

commit c2a3ff1a08df8619f7af77ebe6c3b7661e858bc1
Author: João Barbosa <joao@uphold.com>
Date:   Wed Mar 30 02:04:22 2016 +0100

commit 699720b3baa9e0d3dab679b1da91ab45808aa85e
Author: João Barbosa <joao@uphold.com>
Date:   Wed Mar 30 00:59:29 2016 +0100
```

File: src/wallet/wallet.cpp [hunk 21]
```diff
@@ -2159,14 +2169,24 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
...
                     {
-                        // Insert change txn at random position:
-                        nChangePosRet = GetRandInt(txNew.vout.size()+1);
-                        vector<CTxOut>::iterator position = txNew.vout.begin()+nChangePosRet;
+                        if (nChangePosInOut == -1)
+                        {
+                            // Insert change txn at random position:
+                            nChangePosInOut = GetRandInt(txNew.vout.size()+1);
+                        }
+                        else if (nChangePosInOut > txNew.vout.size())
```
Comment (promag) [inline src/wallet/wallet.cpp]: @jonasschnelli I think the best place to check if index is out of range is here.

**Comment by jonasschnelli** [comment]
Tested ACK 18be394cd8cf5b021e25fff34066f1a3a002858c.

This is very useful.
It allows to fully decouple the private keys from the node/wallet.

**Comment by btcdrak** [comment]
utACK 18be394

**Comment by laanwj** [comment]
Merged this as c0e7d45, with a trivial rebase (only overlap in test framework) to master
- 41e835d (was https://github.com/bitcoin/bitcoin/pull/7518/commits/f7dc1d32bbd23ff593e63f575a2d446a96e1ea15) Add strict flag to RPCTypeCheckObj 
- af4fe7f (was https://github.com/bitcoin/bitcoin/pull/7518/commits/db992eadbcbde0ca14df66c120a5868938e61e83) Add change options to fundrawtransaction 
- f2d0944 (was https://github.com/bitcoin/bitcoin/pull/7518/commits/18be394cd8cf5b021e25fff34066f1a3a002858c) Add lockUnspents option to fundrawtransaction
